File: android/app/src/test/java/com/circuvent/hrms/domain/ShiftRulesTest.kt
package com.circuvent.hrms.domain

import com.circuvent.hrms.domain.ShiftRules.Shift
import com.circuvent.hrms.domain.ShiftRules.State
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime

class ShiftRulesTest {

    private fun shift(
        id: String = "a",
        shiftDate: String = "2026-03-10",
        startsAt: String = "2026-03-10T09:00:00Z",
        endsAt: String = "2026-03-10T17:00:00Z",
        durationMinutes: Int = 480,
        patternName: String? = "Day shift",
    ) = Shift(id, shiftDate, startsAt, endsAt, durationMinutes, "scheduled", patternName)

    private val before = Instant.parse("2026-03-10T06:00:00Z")
    private val during = Instant.parse("2026-03-10T12:00:00Z")
    private val after = Instant.parse("2026-03-10T20:00:00Z")

    @Test
    fun `state follows the clock`() {
        assertEquals(State.UPCOMING, ShiftRules.stateOf(shift(), before))
        assertEquals(State.IN_PROGRESS, ShiftRules.stateOf(shift(), during))
        assertEquals(State.PAST, ShiftRules.stateOf(shift(), after))
    }

    @Test
    fun `the boundaries belong to the shift that is running`() {
        // Not "upcoming" one millisecond after it began, and somebody whose
        // shift ended exactly now is not still on it.
        assertEquals(State.IN_PROGRESS, ShiftRules.stateOf(shift(), Instant.parse("2026-03-10T09:00:00Z")))
        assertEquals(State.PAST, ShiftRules.stateOf(shift(), Instant.parse("2026-03-10T17:00:00Z")))
    }

    @Test
    fun `an unreadable timestamp degrades to past rather than throwing`() {
        assertEquals(State.PAST, ShiftRules.stateOf(shift(startsAt = "not a date"), during))
        assertEquals(State.PAST, ShiftRules.stateOf(shift(endsAt = ""), during))
    }

    @Test
    fun `an offset timestamp is accepted as well as a Z`() {
        // Not every server sends Z. Rejecting an offset would silently drop
        // every shift for anyone whose backend serialises with one.
        assertEquals(
            State.IN_PROGRESS,
            ShiftRules.stateOf(
                shift(startsAt = "2026-03-10T14:30:00+05:30", endsAt = "2026-03-10T22:30:00+05:30"),
                during,
            ),
        )
    }

    @Test
    fun `nothing ahead means nothing to show`() {
        assertNull(ShiftRules.next(emptyList(), during))
        assertNull(ShiftRules.next(listOf(shift()), after))
    }

    @Test
    fun `the shift being worked wins over the one that follows`() {
        val running = shift(id = "running")
        val later = shift(
            id = "later",
            shiftDate = "2026-03-11",
            startsAt = "2026-03-11T09:00:00Z",
            endsAt = "2026-03-11T17:00:00Z",
        )
        assertEquals("running", ShiftRules.next(listOf(later, running), during)?.id)
    }

    @Test
    fun `the earliest upcoming shift wins, whatever the input order`() {
        val soon = shift(id = "soon")
        val later = shift(
            id = "later",
            shiftDate = "2026-03-12",
            startsAt = "2026-03-12T09:00:00Z",
            endsAt = "2026-03-12T17:00:00Z",
        )
        assertEquals("soon", ShiftRules.next(listOf(later, soon), before)?.id)
        assertEquals("soon", ShiftRules.next(listOf(soon, later), before)?.id)
    }

    @Test
    fun `of two overlapping live shifts the most recently started is the one being worked`() {
        val early = shift(id = "early", startsAt = "2026-03-10T08:00:00Z")
        val late = shift(id = "late", startsAt = "2026-03-10T11:00:00Z")
        assertEquals("late", ShiftRules.next(listOf(early, late), during)?.id)
    }

    @Test
    fun `days are grouped on the rostered date, earliest first`() {
        val days = ShiftRules.groupByDay(
            listOf(shift(id = "b", shiftDate = "2026-03-12"), shift(id = "a", shiftDate = "2026-03-10")),
        )
        assertEquals(listOf("2026-03-10", "2026-03-12"), days.map { it.date })
    }

    @Test
    fun `a night shift stays on the day it was rostered for`() {
        // The whole reason grouping uses shiftDate. This starts at 22:00 on the
        // 10th and ends on the 11th; the person was told to come in on the
        // 10th, so that is the day it belongs to. Deriving the day from the
        // start instant puts half a night rota on one date and half on another.
        val night = shift(
            shiftDate = "2026-03-10",
            startsAt = "2026-03-10T22:00:00Z",
            endsAt = "2026-03-11T06:00:00Z",
        )
        assertEquals(listOf("2026-03-10"), ShiftRules.groupByDay(listOf(night)).map { it.date })
    }

    @Test
    fun `a day's shifts are ordered by when they start, and the input is not mutated`() {
        val evening = shift(id = "evening", startsAt = "2026-03-10T18:00:00Z")
        val morning = shift(id = "morning", startsAt = "2026-03-10T06:00:00Z")
        val input = listOf(evening, morning)

        val day = ShiftRules.groupByDay(input).first()
        assertEquals(listOf("morning", "evening"), day.shifts.map { it.id })
        assertEquals(listOf("evening", "morning"), input.map { it.id })
    }

    @Test
    fun `a day totals its minutes`() {
        val days = ShiftRules.groupByDay(
            listOf(
                shift(id = "a", durationMinutes = 240),
                shift(id = "b", startsAt = "2026-03-10T14:00:00Z", durationMinutes = 180),
            ),
        )
        assertEquals(420, days.first().totalMinutes)
    }

    @Test
    fun `overnight is decided in the zone the worker lives in`() {
        // Built from local time on purpose. A fixture written as a UTC instant
        // answers a different question and passes or fails depending on where
        // the test is run: in India, 22:00Z and 06:00Z the next day are the
        // same local date.
        val zone = ZoneId.systemDefault()
        fun local(y: Int, m: Int, d: Int, h: Int): String =
            ZonedDateTime.of(y, m, d, h, 0, 0, 0, zone).toInstant().toString()

        assertFalse(
            ShiftRules.isOvernight(shift(startsAt = local(2026, 3, 10, 9), endsAt = local(2026, 3, 10, 17)), zone),
        )
        assertTrue(
            ShiftRules.isOvernight(shift(startsAt = local(2026, 3, 10, 22), endsAt = local(2026, 3, 11, 6)), zone),
        )
        // Across a month and a year boundary, where comparing day-of-month
        // alone would call it a same-day shift.
        assertTrue(
            ShiftRules.isOvernight(shift(startsAt = local(2026, 3, 31, 22), endsAt = local(2026, 4, 1, 6)), zone),
        )
        assertTrue(
            ShiftRules.isOvernight(shift(startsAt = local(2026, 12, 31, 22), endsAt = local(2027, 1, 1, 6)), zone),
        )
    }

    @Test
    fun `an unreadable timestamp is not reported as overnight`() {
        // Claiming a shift runs overnight on the strength of a value that could
        // not be parsed would tell somebody to arrange a night off they do not
        // need.
        assertFalse(ShiftRules.isOvernight(shift(endsAt = "nonsense")))
    }

    @Test
    fun `durations read the way a person would say them`() {
        assertEquals("0m", ShiftRules.formatDuration(0))
        assertEquals("45m", ShiftRules.formatDuration(45))
        assertEquals("1h", ShiftRules.formatDuration(60))
        assertEquals("8h", ShiftRules.formatDuration(480))
        assertEquals("1h 30m", ShiftRules.formatDuration(90))
        assertEquals("8h 30m", ShiftRules.formatDuration(510))
    }

    @Test
    fun `a negative duration is shown as absent rather than as a negative`() {
        assertEquals("—", ShiftRules.formatDuration(-30))
    }

    @Test
    fun `today, tomorrow and yesterday are named`() {
        val today = LocalDate.of(2026, 3, 10)
        assertEquals("Today", ShiftRules.dayLabel("2026-03-10", today))
        assertEquals("Tomorrow", ShiftRules.dayLabel("2026-03-11", today))
        assertEquals("Yesterday", ShiftRules.dayLabel("2026-03-09", today))
    }

    @Test
    fun `any other day is written out with the right weekday`() {
        // 2026-03-14 is a Saturday. Formatting a date parsed at UTC midnight in
        // the device zone reports Friday for anyone west of Greenwich; a
        // LocalDate has no such trap, and this pins it.
        val label = ShiftRules.dayLabel("2026-03-14", LocalDate.of(2026, 3, 10), java.util.Locale.UK)
        assertTrue("label was: $label", label.contains("Sat"))
        assertTrue(label.contains("14"))
    }

    @Test
    fun `anything that is not a date passes straight through`() {
        assertEquals("soon", ShiftRules.dayLabel("soon", LocalDate.of(2026, 3, 10)))
    }

    @Test
    fun `an unreadable instant does not render as a broken time`() {
        assertEquals("—", ShiftRules.formatClock(""))
        assertEquals("—", ShiftRules.formatClock("nonsense"))
    }
}

File: mobile/src/lib/shift-rules.test.ts
import { describe, expect, it } from "vitest";
import {
  addDays,
  dayLabel,
  formatClock,
  formatDuration,
  groupByDay,
  isOvernight,
  nextShift,
  shiftState,
  type ShiftAssignment,
} from "./shift-rules";

function shift(overrides: Partial<ShiftAssignment> = {}): ShiftAssignment {
  return {
    id: "a",
    shiftDate: "2026-03-10",
    startsAt: "2026-03-10T09:00:00.000Z",
    endsAt: "2026-03-10T17:00:00.000Z",
    durationMinutes: 480,
    status: "scheduled",
    ...overrides,
  };
}

const DURING = new Date("2026-03-10T12:00:00.000Z");
const BEFORE = new Date("2026-03-10T06:00:00.000Z");
const AFTER = new Date("2026-03-10T20:00:00.000Z");

describe("shiftState", () => {
  it("reports a shift that has not started as upcoming", () => {
    expect(shiftState(shift(), BEFORE)).toBe("upcoming");
  });

  it("reports a shift being worked as in progress", () => {
    expect(shiftState(shift(), DURING)).toBe("in_progress");
  });

  it("reports a finished shift as past", () => {
    expect(shiftState(shift(), AFTER)).toBe("past");
  });

  it("treats the start instant as in progress and the end instant as past", () => {
    // A shift is not "upcoming" one millisecond after it began, and someone
    // whose shift ended exactly now is not still on it.
    expect(shiftState(shift(), new Date("2026-03-10T09:00:00.000Z"))).toBe("in_progress");
    expect(shiftState(shift(), new Date("2026-03-10T17:00:00.000Z"))).toBe("past");
  });

  it("degrades to past rather than throwing on an unparseable timestamp", () => {
    expect(shiftState(shift({ startsAt: "not a date" }), DURING)).toBe("past");
    expect(shiftState(shift({ endsAt: "" }), DURING)).toBe("past");
  });
});

describe("nextShift", () => {
  it("returns nothing for an empty roster", () => {
    expect(nextShift([], DURING)).toBeUndefined();
  });

  it("returns nothing when every shift is behind us", () => {
    expect(nextShift([shift()], AFTER)).toBeUndefined();
  });

  it("prefers the shift being worked over the one that follows it", () => {
    const running = shift({ id: "running" });
    const later = shift({
      id: "later",
      shiftDate: "2026-03-11",
      startsAt: "2026-03-11T09:00:00.000Z",
      endsAt: "2026-03-11T17:00:00.000Z",
    });

    expect(nextShift([later, running], DURING)?.id).toBe("running");
  });

  it("picks the earliest upcoming shift regardless of input order", () => {
    const soon = shift({ id: "soon", startsAt: "2026-03-10T09:00:00.000Z" });
    const later = shift({
      id: "later",
      shiftDate: "2026-03-12",
      startsAt: "2026-03-12T09:00:00.000Z",
      endsAt: "2026-03-12T17:00:00.000Z",
    });

    expect(nextShift([later, soon], BEFORE)?.id).toBe("soon");
    expect(nextShift([soon, later], BEFORE)?.id).toBe("soon");
  });

  it("picks the most recently started of two overlapping live shifts", () => {
    const early = shift({ id: "early", startsAt: "2026-03-10T08:00:00.000Z" });
    const late = shift({ id: "late", startsAt: "2026-03-10T11:00:00.000Z" });

    expect(nextShift([early, late], DURING)?.id).toBe("late");
  });

  it("skips a shift whose start cannot be read instead of returning it", () => {
    const broken = shift({ id: "broken", startsAt: "", endsAt: "" });
    const good = shift({ id: "good" });

    expect(nextShift([broken, good], BEFORE)?.id).toBe("good");
  });
});

describe("groupByDay", () => {
  it("returns nothing for an empty roster", () => {
    expect(groupByDay([])).toEqual([]);
  });

  it("groups on the rostered date, earliest day first", () => {
    const days = groupByDay([
      shift({ id: "b", shiftDate: "2026-03-12" }),
      shift({ id: "a", shiftDate: "2026-03-10" }),
    ]);

    expect(days.map((d) => d.date)).toEqual(["2026-03-10", "2026-03-12"]);
  });

  it("orders a day's shifts by when they start", () => {
    const evening = shift({ id: "evening", startsAt: "2026-03-10T18:00:00.000Z" });
    const morning = shift({ id: "morning", startsAt: "2026-03-10T06:00:00.000Z" });

    const [day] = groupByDay([evening, morning]);
    expect(day?.shifts.map((s) => s.id)).toEqual(["morning", "evening"]);
  });

  it("keeps a night shift on the day it was rostered for", () => {
    // The whole reason grouping uses shiftDate. This shift starts at 22:00 on
    // the 10th and ends on the 11th; the person was told to come in on the
    // 10th, so that is the day it belongs to.
    const night = shift({
      shiftDate: "2026-03-10",
      startsAt: "2026-03-10T22:00:00.000Z",
      endsAt: "2026-03-11T06:00:00.000Z",
    });

    expect(groupByDay([night]).map((d) => d.date)).toEqual(["2026-03-10"]);
  });

  it("totals the minutes on each day", () => {
    const days = groupByDay([
      shift({ id: "a", durationMinutes: 240 }),
      shift({ id: "b", durationMinutes: 180, startsAt: "2026-03-10T14:00:00.000Z" }),
    ]);

    expect(days[0]?.totalMinutes).toBe(420);
  });

  it("does not mutate the roster it was given", () => {
    const evening = shift({ id: "evening", startsAt: "2026-03-10T18:00:00.000Z" });
    const morning = shift({ id: "morning", startsAt: "2026-03-10T06:00:00.000Z" });
    const input = [evening, morning];

    groupByDay(input);

    expect(input.map((s) => s.id)).toEqual(["evening", "morning"]);
  });
});

describe("isOvernight", () => {
  /**
   * An instant expressed in the runner's own timezone.
   *
   * Built this way deliberately. `isOvernight` asks the question the person on
   * the shift would ask — do I go home tomorrow — so it compares local
   * calendar days. A fixture written as a UTC instant answers a different
   * question and passes or fails depending on where the test is run: in
   * India, 22:00Z and 06:00Z the next day are both the same local date.
   */
  function localIso(year: number, month: number, day: number, hour: number): string {
    return new Date(year, month - 1, day, hour, 0, 0, 0).toISOString();
  }

  it("is false for a shift that starts and ends on one day", () => {
    expect(isOvernight(shift())).toBe(false);
    expect(
      isOvernight(
        shift({ startsAt: localIso(2026, 3, 10, 9), endsAt: localIso(2026, 3, 10, 17) })
      )
    ).toBe(false);
  });

  it("is true when the end falls on a later local day", () => {
    expect(
      isOvernight(
        shift({ startsAt: localIso(2026, 3, 10, 22), endsAt: localIso(2026, 3, 11, 6) })
      )
    ).toBe(true);
  });

  it("is true across a month boundary", () => {
    expect(
      isOvernight(
        shift({ startsAt: localIso(2026, 3, 31, 22), endsAt: localIso(2026, 4, 1, 6) })
      )
    ).toBe(true);
  });

  it("is true across a year boundary", () => {
    // Comparing day-of-month alone would call this a same-day shift.
    expect(
      isOvernight(
        shift({ startsAt: localIso(2026, 12, 31, 22), endsAt: localIso(2027, 1, 1, 6) })
      )
    ).toBe(true);
  });

  it("is false rather than true when a timestamp cannot be read", () => {
    // Claiming a shift runs overnight on the strength of a value that could
    // not be parsed would tell someone to arrange a night off they do not need.
    expect(isOvernight(shift({ endsAt: "nonsense" }))).toBe(false);
    expect(isOvernight(shift({ startsAt: "" }))).toBe(false);
  });
});

describe("formatDuration", () => {
  it("writes minutes under an hour on their own", () => {
    expect(formatDuration(0)).toBe("0m");
    expect(formatDuration(45)).toBe("45m");
  });

  it("drops the minutes when there are none", () => {
    expect(formatDuration(60)).toBe("1h");
    expect(formatDuration(480)).toBe("8h");
  });

  it("writes hours and minutes together", () => {
    expect(formatDuration(90)).toBe("1h 30m");
    expect(formatDuration(510)).toBe("8h 30m");
  });

  it("refuses to render a negative or non-finite duration as a number", () => {
    expect(formatDuration(-30)).toBe("—");
    expect(formatDuration(Number.NaN)).toBe("—");
    expect(formatDuration(Number.POSITIVE_INFINITY)).toBe("—");
  });
});

describe("addDays", () => {
  it("moves forward and back across a month boundary", () => {
    expect(addDays("2026-03-31", 1)).toBe("2026-04-01");
    expect(addDays("2026-03-01", -1)).toBe("2026-02-28");
  });

  it("handles a leap day", () => {
    expect(addDays("2028-02-28", 1)).toBe("2028-02-29");
  });

  it("returns the input unchanged when it is not a date", () => {
    expect(addDays("later", 1)).toBe("later");
  });
});

describe("dayLabel", () => {
  const today = "2026-03-10";

  it("names today, tomorrow and yesterday", () => {
    expect(dayLabel("2026-03-10", today)).toBe("Today");
    expect(dayLabel("2026-03-11", today)).toBe("Tomorrow");
    expect(dayLabel("2026-03-09", today)).toBe("Yesterday");
  });

  it("writes any other day out", () => {
    const label = dayLabel("2026-03-14", today);
    expect(label).not.toBe("Today");
    expect(label).toContain("14");
  });

  it("names the weekday from the calendar date, not the device timezone", () => {
    // 2026-03-14 is a Saturday. Formatting it in the device zone after parsing
    // it at UTC midnight reports Friday for anyone west of Greenwich.
    expect(dayLabel("2026-03-14", today)).toMatch(/Sat/);
  });

  it("passes through anything that is not a date", () => {
    expect(dayLabel("soon", today)).toBe("soon");
  });
});

describe("formatClock", () => {
  it("refuses an unreadable instant rather than printing Invalid Date", () => {
    expect(formatClock("")).toBe("—");
    expect(formatClock("nonsense")).toBe("—");
  });

  it("renders a real instant as a time", () => {
    expect(formatClock("2026-03-10T09:00:00.000Z")).toMatch(/\d/);
  });
});
